HIP/hip-1056.md

```diff
@@ -0,0 +1,1769 @@
+- **Reconnect Services**: Reconnect enables nodes that have fallen behind other nodes on the network to request data
+  from nodes that are in sync. This process requires significant complexity and resources. We can enhance stability and
+  simplify node software by allowing nodes to reconnect with a block node, instead of another consensus node. This
+  is a significant and important enhancement to reconnect that is needed for permissionless nodes.
+- **Query Endpoints:** As Block Nodes or other consumers of the Block Stream can maintain a live state in sync with the
+  consensus nodes, they can answer any queries against state. They can also produce cryptograph state proofs to prove
+  any data in state (a `state proof`) or any data was part of a block (a `block proof`). For example, they can provide a
+  `state proof` that says a particular account has a specific balance, or they can provide a `block proof` that a
+  given HCS message was part of a specific block.
+
+## Rationale
+
+As mentioned in the motivation section above, the design rationale of Block Streams is driven by the following goals:
+
+**Consolidation of Streams**: Block streams should efficiently consolidate all information across event, state, and
+record streams, optimizing data size to reduce network storage costs over time.
+
+**Verifiability**: Each block must be self-contained and independently verifiable. Block streams will provide the data
+to enable verification of hashgraph consensus as well as verification of transaction execution.
+
+**State Update Data:** Block streams will include all state changes to enable downstream services to maintain merkle
+state alongside mainnet nodes. Prior to block streams the contents of consensus state was not directly visible from
+outside the consensus network.
+
+**Easy consumption**: The Block stream is designed to be easy to consume by users. Where trade-offs exist between a
+small amount of extra work on consensus nodes or a large amount of work or complex understanding needed on consumers,
+we have chosen to perform extra work on consensus nodes to produce clean block streams.
+
+**Designed for Streaming or as Block files:** The stream is designed to both be packaged up as self-contained block
+files and to be streamed as individual items with low latency.
+
+#### Design for Consolidation
+
+The block stream is designed to be a continuous stream of items. This allows for continuous streaming over gRPC. The
+order of items is significant, for example all items between and inclusive of a `BlockHeader` and `BlockProof` represent
+contents of one block. All items following an `EventTransaction` before the next `EventTransaction` are related to that
+first transaction. Block items for one block can also be packaged in a `Block` message so they can be stored in a
+protobuf file and served as a single entity if desired.
+
+![Block Stream Items](../assets/hip-blocks/block-stream-items.svg)
```

Comment:
There are some cases where multiple changes made to a singleton state type over the course of a round will be consolidated into a single change set for that singleton type for the round instead. How do those fit into the stream and block merkle tree? Do they need to be placed somewhere specific both in the stream and in the tree?

Reply:
It all depends on how that affects the merkle tree in consensus state.
As long as we end up with the same state hash in the block node as the consensus node, the state change can be placed after pretty much any transaction.
I _think_ services currently puts all singleton updates, at the end of the block (after the last transaction, but before the block proof).